marketdata/stream/client_test.go

```diff
@@ -207,6 +207,52 @@ func TestConnectSucceeds(t *testing.T) {
 	}
 }
 
+func TestCallbacksCalledOnConnectAndDisconnect(t *testing.T) {
+	for _, tt := range tests {
+		numConnectCalls := 0
+		numDisconnectCalls := 0
+		connectCallback := func() { numConnectCalls++ }
+		disconnectCallback := func() { numDisconnectCalls++ }
+		t.Run(tt.name, func(t *testing.T) {
+			connection := newMockConn()
+			defer connection.close()
+			connCreator := func(ctx context.Context, u url.URL) (conn, error) {
+				return connection, nil
+			}
+
+			writeInitialFlowMessagesToConn(t, connection, subscriptions{})
+
+			var c StreamClient
+			switch tt.name {
+			case stocksTests:
+				c = NewStocksClient("iex",
+					withConnCreator(connCreator),
+					WithConnectCallback(connectCallback),
+					WithDisconnectCallback(disconnectCallback))
+			case cryptoTests:
+				c = NewCryptoClient(withConnCreator(connCreator),
+					WithConnectCallback(connectCallback),
+					WithDisconnectCallback(disconnectCallback))
+			}
+
+			ctx, cancel := context.WithCancel(context.Background())
+			defer cancel()
+
+			err := c.Connect(ctx)
+			require.NoError(t, err)
+			assert.Equal(t, 1, numConnectCalls)
+			assert.Equal(t, 0, numDisconnectCalls)
+
+			// Now force the stream to disconnect via context and assert disconnect is called after
+			// waiting a small amount of time to wait for the stream to shut down
+			cancel()
+			time.Sleep(100 * time.Millisecond)
```

Comment:
Please do not use sleep in unit tests. Here's how you should do this instead: in the disconnect callback function you should write to a channel, and here you should read from that channel with a timeout:
``` go
select {
case <-disconnects:
case <-time.After(time.Second):
	require.Fail(t, "disconnect callback was not called in time")
}
```